share/doc/wake/wake-for-fsharp-developers.md

```diff
@@ -0,0 +1,150 @@
+
+```fsharp
+let numberTwo = 2
+
+let rec fibonacci (n: int) : int =
+    match n with
+    | 0 | 1 -> n
+    | n -> fibonacci (n-1) + fibonacci (n-2)
+```
+
+In Wake, bindings are created with `def` and do not require any keyword if they are recursive.
+
+```wake
+def numberTwo = 2
+
+def fibonacci (n: Integer): Integer =
+    match n
+        0 = n
+        1 = n
+        n = fibonacci (n-1) + fibonacci (n-2)
+```
+
+## Lists
+
+In F#, cons is `::` and the empty list is `[]`. There is literal list syntax with `[]` and `;` as a separator.
+
+```fsharp
+> 1 :: 2 :: 3 :: [];;
+val it: int list = [1; 2; 3]
+
+> [1; 2; 3];;
+val it: int list = [1; 2; 3]
+```
+
+In Wake, cons is [`,`](https://github.com/sifive/wake/blob/master/share/wake/lib/core/list.wake#L43-L50) and the empty list is [`Nil`](https://github.com/sifive/wake/blob/master/share/wake/lib/core/list.wake#L41-L42). There is currently no literal list syntax in Wake. You often need to surround a list value with parentheses, such as `(1, 2, 3, Nil)`.
+
+```wake
+> wake -vx '1, 2, 3, Nil'
+1, 2, 3, Nil: List Integer = 1, 2, 3, Nil
+```
+
+## Discriminated unions vs `data` types
+
+F#
+```fsharp
+type Shape =
+    | Circle of radius: float
+    | Square of side: float
+    | Rectangle of length: float * width: float
+```
+
+Wake
+```wake
+export data Shape =
+    Circle (radius: Double)
+    Square (side: Double)
+    Rectangle (length: Double) (width: Double)
+```
+
+## Pattern Matching
+
+F#
+```fsharp
+let rec area shape =
+    match shape with
+    | Circle radius             -> radius * radius
+    | Square side               -> area (Rectangle (side, side))
+    | Rectangle (length, width) -> length * width
+```
+
+Wake
+```wake
+export def area (shape: Shape) =
+    match shape
+        Circle radius          = pi *. (radius ^. 2.0)
+        Square side            = area (Rectangle side side)
+        Rectangle length width = length *. width
+```
+
+Note that in F#, recursive functions require the `rec` keyword, whereas in Wake, recursive functions require no special keyword.
+
+## Piping
+
+F#
+```fsharp
+let square (x: int) = x * x
+
+2 |> square
+```
+
+Wake
+```wake
+export def square (x: Integer) = x * x
+
+2 | square
+```
+
+## Records, tuples, pairs etc.
+
+Records in F# are analogous to tuples in Wake.
+
+F#
+```fsharp
+// This is a record
+type Name = {First: string; Last: string}
+
+// This is a value of type `Name`
+{First: "John"; Last: "Smith"}
+
+// This is a record with a generic type parameter
+type Person<'T> = {Name: Name; JobFunction: 'T}
+```
+
+Wake
+```wake
+# This is a tuple
+export tuple Name =
+    First: String
+    Last: String
+
+# This a value of type `Name`
+export def name = Name "John" "Smith"
+
+# This is a tuple with a type parameter
+export tuple Person a =
+    Name: Name
+    JobFunction: a
+
+# This is a value of type `Person Integer`
+export def person = Person (Name "John" "Smith") 4
```

Comment:
These two may or may not want the same changes (it would be nice to provide the sleaker not-export-every-line tuple format).

Reply:
That's a good point, but I already added it from above. Haha. I think I'll leave it for now with the `export`s added it for consistency. It might make sense to later update this section to point to some more general Wake documentation.